### ISSUE TITLE
Plots are unreadable when the number of instance constraints is large

### ISSUE THREAD

**Issue by Peter230655** [csu-hmc/opty#213]
If there are many instance constraints, the labels on the x axis overlap each other.
If I understood the code of the function plot_constraint_violations(solution, axis) correctly, it would suffice to give one more optional keyword=rotation to rotate the labels as needed. 

![image](https://github.com/user-attachments/assets/cf89358d-0cd5-452c-a206-049cd5449532)

I could try to make a PR myself, but I cannot find the right file. When I look what is in opty, I get this:
![image](https://github.com/user-attachments/assets/dd2f39b1-35a5-44b9-959a-3d469b3f2e26)

**Comment by moorepants** [comment]
You could but I don't think that will really solve things in the long run. I've tried the rotation before. It would be better to add more rows to the plots so that we have only like max of 10 instances constraints per row or something.

**Comment by Peter230655** [comment]
I kind of figured my idea would not get me a Nobel price. :-)
Do you mean to have more plots, and only, say, 10 bars per plot?

**Comment by moorepants** [comment]
Yes. I'm sure you could add that.

**Comment by moorepants** [comment]
Rotation is a fine idea, I just happened to spend an hour or so on it some time in the past and recall concluding the rotation only solved the problem sometimes (depended both on # of constraints and length of the constraint string).

**Comment by tjstienstra** [comment]
The main problem in my opinion on rotating is that the equations are in fact just too long to write down below the plot. An option would be to just show the indices on the x-axis. Users can then either look at their code for their instance constraints or use some display function.

**Comment by moorepants** [comment]
Yes, that is also a possible fix, but we can still get large numbers of them, so maybe it is a partial fix.

One other thing that makes them long is that floats will print many decimal places. We could print those using scientific notation with a limited number of places or something.

**Comment by Peter230655** [comment]
If I see it correctly, the labels on the x axis are the state variable occurring in the instance constraints and the time. I get e.g.
xB(0), xB(249*h). So, if h is to be minimized, you do not the the 'real' value, and otherwise you know at which points in time the constraints are acting. It does not seem important to get these times very accurately - or am I seeing something wrong?

**Comment by moorepants** [comment]
The purpose of the labels is to easily know what constraint is associated with what bar. We currently call `str()` on the constraint expression. The default behavior displays a large number (non-)significant figures.

**Comment by Peter230655** [comment]
So a few significant digits would suffice? Is someone wants to know exactly, they can always look at info['g'] ?
On the EoM violation it does not say at all, which state variable has which error. Again for space reasons?

**Comment by moorepants** [comment]
>  On the EoM violation it does not say at all, which state variable has which error. Again for space reasons?

The violation isn't associated with a state, it is associated with an equation of motion and those are almost always too long to print.

**Comment by Peter230655** [comment]
> > On the EoM violation it does not say at all, which state variable has which error. Again for space reasons?
> 
> The violation isn't associated with a state, it is associated with an equation of motion and those are almost always too long to print.

So it | eom[i] |, as ideally eom[i] = 0 for all i (?)
I had wondered how it determines | $y  - y_{true} | since $y_{true}$ is unknown, else no need to calculate it. Now I understand.

**Comment by Peter230655** [comment]
I'd like to play around with this a bit to have something to do (I am playing around with one more example for examples-gallery, not yet completely finished, but I am running out of new ideas..).
As I wrote above, I cannot find the file or directory it is in.
(When I messed around with KanesMethod, I found _kane_ in sympy/physics/mechanics - but where do I find plot_constraints_violations?
Thanks!

**Comment by moorepants** [comment]
If you'd like a challenging example to get work, have a look at this one: https://github.com/csu-hmc/opty/pull/29. I haven't had time to get it working (for a decade now!).

**Comment by moorepants** [comment]
>  (When I messed around with KanesMethod, I found kane in sympy/physics/mechanics - but where do I find plot_constraints_violations?

Github search and command line grep are your friends.

**Comment by Peter230655** [comment]
> If you'd like a challenging example to get work, have a look at this one: #29. I haven't had time to get it working (for a decade now!).



> > (When I messed around with KanesMethod, I found kane in sympy/physics/mechanics - but where do I find plot_constraints_violations?
> 
> Github search and command line grep are your friends.

More my enimies! I found the code in Github, that is where I saw _rotation_.
It seems to be in a class, which I could not find in opty. What I mean is this:
with git I should find a path: cd opty/..... and I cannot.

**Comment by Peter230655** [comment]
> If you'd like a challenging example to get work, have a look at this one: #29. I haven't had time to get it working (for a decade now!).

**Sounds interesting!** I found the code and some .yml file with the constants. I am usure what.yml is, but I can look it up.
Is there any paper decribing what it does or just try and get the code to run?
(I had seen this old 'thing' often and wondered what it was...)

**Comment by moorepants** [comment]
>  Is there any paper decribing what it does or just try and get the code to run?

Let's move the discussion to that issue.

**Comment by moorepants** [comment]
Example from the one legged cycling example:

![image](https://github.com/user-attachments/assets/d08f9dfe-7d3c-47a5-8a21-778f8d7a1df3)

Note the length of the instance constraint equations (notably floating point display) and that there can be overlap. This plot has a 12 inch wide figure to try to accommodate but it still has some display issues.

**Comment by Peter230655** [comment]
I am playing around with this right now.
My approach is this:

- optional key_word bars_per_plot, default is, say, 8
- if len(instance_constraints) > bars_per_plot, there will be more plots with the instance violations below each other.
- optional key_word rotation, default is rotation=-10

Should I pursue this approach?

**Comment by moorepants** [comment]
I would try to solve it first without adding any optional keyword arguments. It's better if it just works. If someone wants to tweak a plot they can always make their own with matplotlib.

**Comment by Peter230655** [comment]
> I would try to solve it first without adding any optional keyword arguments. It's better if it just works. If someone wants to tweak a plot they can always make their own with matplotlib.

No problem, then bars_per_plot and rotation are fixed.
Rest of the approach acceptable?

**Comment by moorepants** [comment]
Yes, I'm fine with adding more plots as a function of the number of constraints.

**Comment by Peter230655** [comment]
Thanks, I will continue to try.
One question to optional key_word: A user does not have to do anything if he does not want or does not know about them?

I will of course test it with one of my simulations. Should I manage to get it right and make a PR do I have to supply a test?
After all, any opty simulation would be a test for it.

**Comment by moorepants** [comment]
That's true but matplotlib plots have hundreds of options and it is a fools errand to propagate those options up into our API. The purpose of these plotting methods on Problem are to have a quick basic look at the solution. If you want to make a plot for publication or other purposes you should manually construct those. For these plot methods on Problem we should keep them simple and try to avoid having lots of plot customization options. You can always modify the axes that it returns anyway you like after the default plot is made.

**Comment by moorepants** [comment]
> I will of course test it with one of my simulations. Should I manage to get it right and make a PR do I have to supply a test?

Test it with the entire example gallery and any other more extreme cases that you have on hand. And yes, then make a PR.

**Comment by Peter230655** [comment]
> That's true but matplotlib plots have hundreds of options and it is a fool's errand to propagate those options up into our API. The purpose of these plotting methods on Problem are to have a quick basic look at the solution. If you want to make a plot for publication or other purposes you should manually construct those. For these plot methods on Problem we should keep them simple and try to avoid having lots of plot customization options. You can always modify the axes that it returns anyway you like after the default plot is made.

True! Actually this is how I did it before I heard about this method from you!

**Comment by Peter230655** [comment]
This is what it looks lik right now. Anychanges you suggest?
I allow 10 bars per plot, the rotation is -20
On purpose I made one constrain with many digits.
I have no simple idea on how to shorten such long numbers. 
![image](https://github.com/user-attachments/assets/73366495-480d-4433-998e-4ce77213d256)

**Comment by moorepants** [comment]
My only suggestion it to reduce the number of numbers after the decimal. It is easiest if you open a PR and we can test it and see the results on the entire example gallery.

**Comment by moorepants** [comment]
You can shorten long numbers using Python's builtin support for printing floats (or you can use SymPy's).

**Comment by Peter230655** [comment]
> You can shorten long numbers using Python's builtin support for printing floats (or you can use SymPy's).

The tick labels seem to be of type <class 'sympy.core.add.Add'>
How would I break them apart in the state_variable and the number?

**Comment by Peter230655** [comment]
> My only suggestion it to reduce the number of numbers after the decimal. It is easiest if you open a PR and we can test it and see the results on the entire example gallery.

Should I first try to get the long numbers right, or open a PR right now?

**Comment by tjstienstra** [comment]
> How would I break them apart in the state_variable and the number?

A really simple solution is shown [hre](https://stackoverflow.com/a/43808582).

**Comment by Peter230655** [comment]
> > How would I break them apart in the state_variable and the number?
> 
> A really simple solution is shown [here](https://stackoverflow.com/a/43808582). A bit better would probably be to use `atoms` to find all floats and do a xreplace at once.

**Thanks!!** I think, I will try and copy the example.

**Comment by Peter230655** [comment]
**works perfect!!!**
Is there a way to have the number in exponential?
I mean like 50.12345 = 5.123e2

**Comment by Peter230655** [comment]
The output would look like this, with 14 instance constraints.
The constraints are plotted to 3 digits (@tjstienstra: THANKS!!)
I have tested it only with one simulation, of course changed around many things.
_Should I make a PR?_
![image](https://github.com/user-attachments/assets/3878431e-56ca-4f37-9c80-d47e2efd0d89)

**Comment by tjstienstra** [comment]
Is `qB(0) - 1.123` correct? In the previous plot you've shown it was using scientific notation with e-15.

**Comment by Peter230655** [comment]
> Is `qB(0) - 1.123` correct? In the previous plot you've shown it was using scientific notation with e-15.

I changed it. I am testing it with one simulation only, so I change around a lot.

**Comment by tjstienstra** [comment]
> Is there a way to have the number in exponential?
> I mean like 50.12345 = 5.123e2

I assume that you mean something like this:
```python
import sympy as sm
from sympy.printing.latex import LatexPrinter
from mpmath.libmp.libmpf import prec_to_dps, to_str as mlib_to_str


class CustomLatexPrinter(LatexPrinter):
    _default_settings = {
        **LatexPrinter._default_settings,
        "max_dps": None,
    }
    def _print_Float(self, expr):
        # Based off of that in StrPrinter
        dps = prec_to_dps(expr._prec)
        if self._settings['max_dps'] is not None:
            dps = min(dps, self._settings['max_dps'])
        strip = False if self._settings['full_prec'] else True
        low = self._settings["min"] if "min" in self._settings else None
        high = self._settings["max"] if "max" in self._settings else None
        str_real = mlib_to_str(expr._mpf_, dps, strip_zeros=strip, min_fixed=low, max_fixed=high)

        # Must always have a mul symbol (as 2.5 10^{20} just looks odd)
        # thus we use the number separator
        separator = self._settings['mul_symbol_latex_numbers']

        if 'e' in str_real:
            (mant, exp) = str_real.split('e')

            if exp[0] == '+':
                exp = exp[1:]
            if self._settings['decimal_separator'] == 'comma':
                mant = mant.replace('.','{,}')

            return r"%s%s10^{%s}" % (mant, separator, exp)
        elif str_real == "+inf":
            return r"\infty"
        elif str_real == "-inf":
            return r"- \infty"
        else:
            if self._settings['decimal_separator'] == 'comma':
                str_real = str_real.replace('.','{,}')
            return str_real

eq = sm.symbols("x") + 50.12345
print(CustomLatexPrinter(settings={"max_dps": 3}).doprint(eq))
```

The other option would a bit of a cheat that boils down to `eq.xreplace({sm.Float(50.12345): sm.Symbol("5.01 \times 10^1"})`.

**Comment by Peter230655** [comment]
**Thanks!**
I think, this is complicated.
As Jason said, this should only be a quick tool, if someone wants better views, they can make their own matplotlib program. He even did not want to add optional keywords for rotation and bars_per_plot.

**Comment by moorepants** [comment]
It doesn't matter how complicated it is internally in opty, my point earlier was only about what the user sees and interacts with.

**Comment by Peter230655** [comment]
> It doesn't matter how complicated it is internally in opty, my point earlier was only about what the user sees and interacts with.

I will try! Between Timo and chatGPT I might be successful! :-)

**Comment by tjstienstra** [comment]
@Peter230655 Note that most of the code of that `CustomLatexPrinter` is just copied from [`LatexPrinter._print_Float`](https://github.com/sympy/sympy/blob/1b08561def04d0e0e87a3868feecd4195a23f9fe/sympy/printing/latex.py#L461).

**Comment by Peter230655** [comment]
![image](https://github.com/user-attachments/assets/626967de-2249-43b9-b568-f2ab88746c22)

**Comment by tjstienstra** [comment]
> ![image](https://private-user-images.githubusercontent.com/83544146/363205145-626967de-2249-43b9-b568-f2ab88746c22.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUwMjc0ODYsIm5iZiI6MTcyNTAyNzE4NiwicGF0aCI6Ii84MzU0NDE0Ni8zNjMyMDUxNDUtNjI2OTY3ZGUtMjI0OS00M2I5LWI1NjgtZjJhYjg4NzQ2YzIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODMwVDE0MTMwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlN2JkNTJjZjYxYTI2YjViYjBhY2QwYmZiMDIwZDk4YTZlOTgyZWM2YjFmZmI0NjA1MjY3YjM1MjkyOTZmYTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bggWlvhxca15Ae5rGSlQgHEAscCsYiqwBxPB-i3bcWE)

Might be nice to only use scientific notation when required, e.g. $-1.23$ instead of $-1.23 \cdot 10^0$

**Comment by Peter230655** [comment]
I will do this, should be easy, even for me! :-)

**Comment by moorepants** [comment]
Fixed in #218
